analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd

```diff
@@ -0,0 +1,293 @@
+  mutate(
+    dominant_cytoband_status = case_when(
+      callable_fraction < (1 - uncallable_threshold) ~ "uncallable",
+      loss_fraction > status_threshold ~ "loss",
+      gain_fraction > status_threshold ~ "gain",
+      loss_fraction + gain_fraction > status_threshold ~ "unstable",
+      TRUE ~ "neutral"
+    )
+  ) 
+
+# Join the consensus seg arm status data and filter to include only neutral
+# chromosome arms and non-neutral cytobands
+filtered_consensus_cytoband_status <- consensus_seg_cytoband_status %>%
+  left_join(
+    consensus_seg_arm_status,
+    by = c(
+      "Kids_First_Biospecimen_ID",
+      "chromosome_arm"
+    )
+  ) %>%
+  # Filter the annotated CN data to include only neutral chromosome arms and disagreements
+  filter(
+    dominant_arm_status %in% c("neutral", "uncallable", "unstable") |
+      (
+        dominant_cytoband_status != dominant_arm_status &
+          # bands that disagree with arm, but are not neutral (or uncallable)
+          !(dominant_cytoband_status %in% c("neutral", "uncallable", "unstable"))
+      )
+  ) %>%
+  select(
+    Kids_First_Biospecimen_ID,
+    cytoband,
+    loss_fraction = loss_fraction.x,
+    gain_fraction = gain_fraction.x,
+    callable_fraction = callable_fraction.x,
+    dominant_cytoband_status
+  )
+
+# Display table
+filtered_consensus_cytoband_status
+```
+
+### Determine gene-level status
+
+```{r}
+# Create separate rows for genes span multiple cytobands
+# CAUTION: This will require addition of a distinct() statment later to resolve duplicates
+# Filtering to only multiband genes first because regexes are slow.
+multi_band_genes <- consensus_seg_autosomes_df %>%
+  filter(grepl("-", cytoband)) %>%
+  extract(cytoband, into = c("chrom", "band"), regex = "([0-9]+)(.+)") %>% # make chrom and band cols
+  separate_rows(band, sep = "-") %>% # duplicate rows with more than one band
+  unite(cytoband, chrom, band, sep = "") # rejoin chrom and band
+
+# Filter to singleband genes
+single_band_genes <- consensus_seg_autosomes_df %>%
+  filter(!grepl("-", cytoband))
+gene_df <- bind_rows(single_band_genes, multi_band_genes)
+
+# Now create a data.frame with the gene-level status calls for the
+# neutral, uncallable, and unstable cytoband-level and chromosome arm-level
+# calls
+filtered_consensus_gene_status <- gene_df %>%
+  left_join(
+    consensus_seg_arm_status,
+    by = c("biospecimen_id" = "Kids_First_Biospecimen_ID",
+           "chromosome_arm")
+  ) %>%
+  left_join(
+    consensus_seg_cytoband_status,
+    by = c("biospecimen_id" = "Kids_First_Biospecimen_ID",
+           "cytoband")
+  ) %>%
+  # Filter the annotated CN data to include only neutral arms and cytobands,
+  # and disagreements
+  filter(
+    dominant_arm_status %in% c("neutral", "uncallable", "unstable") |
+      dominant_cytoband_status %in% c("neutral", "uncallable", "unstable") |
+      status != dominant_arm_status |
+      status != dominant_cytoband_status
```

Comment:
This is the critical bit, and the logic is tough! I think we need some `&`s in here and parenthesis... I'm going to try to figure it out, with annotation.  I would love another set of eyes on this (as many sets as possible?)

```suggestion
  filter(
    # Case 1) Gene call disagrees with both arm and cytoband (This captures most 
    # we want to keep, including all of when arm and cytoband are neutral, uncallable 
    # or unstable, since we have no neutral gene calls in this df):
    (status != dominant_arm_status & status != dominant_cytoband_status)
    # Case 2) Gene call disagrees with a non-neutral cytoband call. 
    # Keep no matter what the arm status
    | (status != dominant_cytoband_status 
       & dominant_cytoband_status %in% c("gain", "loss"))
    # I think that captures everything we want. Cases we don't want include:
    # gene & arm agree, but cytoband is neutral
    # gene & cytoband agree
    # all 3 agree
```

Reply:
Ah, this logic makes much more sense to me, thank you for the added comments here!

Reply:
Those comments were not just for you! They were for me as I was writing it. Not even me tomorrow; me today!

Reply:
Checking my understanding here - we _do_ have neutrals for cytoband and arm. @jashapiro can you tell me a little bit more about
```
# gene & arm agree, but cytoband is neutral
```
as well.

Reply:
Correct, we have neutrals for cytoband and arm. But if we call the whole arm as loss (for example), but the cytoband is called neutral, the assumption I was making is that the cytoband is probably a false negative, and we would already have the arm that "covers" all of the genes in this region, so an individual gene that is lost in that cytoband would not be worth of calling out as "focal".

Reply:
Thanks @jashapiro ! This logic seems correct to me.